### PR TITLE
fix(payouts): persist status updates in payouts table

### PR DESCRIPTION
**Pull request by kashif-m**
refactor(payouts): populate customer fields in list response

## Type of Change

- [x] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
This PR fixes below payout related **BUGS**

- Customer related fields are not populated
- Status is not persisted in `payouts` table

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
This is to be used for listing payouts, and it needs to work as expected!

## How did you test it?
Tested locally using postman.

### List payouts

<details>
  <summary>View request</summary>
  curl --location 'http://localhost:8080/payouts/list?limit=4' \
  --header 'api-key: dev_wmZ4BwC6pXMpBZs3kHlEHAbAxQevfCfJvEESnNmpm7eGlRlNH3yqWCjH58Q87Vfm'
</details>

<details>
  <summary>View response</summary>
  {
    "size": 4,
    "data": [
        {
            "payout_id": "c353a893-a0a6-4be4-849e-a328264b8509",
            "merchant_id": "merchant_1712111780",
            "amount": 1000,
            "currency": "EUR",
            "connector": "wise",
            "payout_type": "bank",
            "billing": null,
            "customer_id": "cus_NOWPAviaOF3z7Ynbi5MQ",
            "auto_fulfill": true,
            "email": "guest@example.com",
            "name": "John Doe",
            "phone": "999999999",
            "phone_country_code": "+65",
            "client_secret": null,
            "return_url": null,
            "business_country": null,
            "business_label": null,
            "description": "Its my first payout request",
            "entity_type": "Individual",
            "recurring": true,
            "metadata": {},
            "status": "success",
            "error_message": null,
            "error_code": null,
            "profile_id": "pro_sxXGxtmZC38hYU6MB624",
            "created": "2024-04-03T05:05:24.754Z",
            "attempts": [
                {
                    "attempt_id": "c353a893-a0a6-4be4-849e-a328264b8509_1",
                    "status": "success",
                    "amount": 1000,
                    "currency": "EUR",
                    "connector": "wise",
                    "error_code": null,
                    "error_message": null,
                    "payment_method": "bank",
                    "payout_method_type": null,
                    "connector_transaction_id": "53200690",
                    "cancellation_reason": null,
                    "unified_code": null,
                    "unified_message": null
                }
            ]
        },
        {
            "payout_id": "c79a5213-7cd7-4d0f-bcfd-d934b69cdf47",
            "merchant_id": "merchant_1712111780",
            "amount": 1000,
            "currency": "EUR",
            "connector": "adyen",
            "payout_type": "bank",
            "billing": null,
            "customer_id": "cus_NOWPAviaOF3z7Ynbi5MQ",
            "auto_fulfill": true,
            "email": "guest@example.com",
            "name": "John Doe",
            "phone": "999999999",
            "phone_country_code": "+65",
            "client_secret": null,
            "return_url": null,
            "business_country": null,
            "business_label": null,
            "description": "Its my first payout request",
            "entity_type": "Individual",
            "recurring": true,
            "metadata": {},
            "status": "success",
            "error_message": null,
            "error_code": null,
            "profile_id": "pro_sxXGxtmZC38hYU6MB624",
            "created": "2024-04-03T05:05:13.365Z",
            "attempts": [
                {
                    "attempt_id": "c79a5213-7cd7-4d0f-bcfd-d934b69cdf47_1",
                    "status": "success",
                    "amount": 1000,
                    "currency": "EUR",
                    "connector": "adyen",
                    "error_code": null,
                    "error_message": null,
                    "payment_method": "bank",
                    "payout_method_type": null,
                    "connector_transaction_id": "NBL39F85X5252VT5",
                    "cancellation_reason": null,
                    "unified_code": null,
                    "unified_message": null
                }
            ]
        },
        {
            "payout_id": "70590cab-c8d5-4109-9a1e-395b59b82ec0",
            "merchant_id": "merchant_1712111780",
            "amount": 1000,
            "currency": "EUR",
            "connector": "adyen",
            "payout_type": "bank",
            "billing": null,
            "customer_id": "cus_NOWPAviaOF3z7Ynbi5MQ",
            "auto_fulfill": true,
            "email": "guest@example.com",
            "name": "John Doe",
            "phone": "999999999",
            "phone_country_code": "+65",
            "client_secret": null,
            "return_url": null,
            "business_country": null,
            "business_label": null,
            "description": "Its my first payout request",
            "entity_type": "Individual",
            "recurring": true,
            "metadata": {},
            "status": "success",
            "error_message": null,
            "error_code": null,
            "profile_id": "pro_sxXGxtmZC38hYU6MB624",
            "created": "2024-04-03T05:05:04.726Z",
            "attempts": [
                {
                    "attempt_id": "70590cab-c8d5-4109-9a1e-395b59b82ec0_1",
                    "status": "success",
                    "amount": 1000,
                    "currency": "EUR",
                    "connector": "adyen",
                    "error_code": null,
                    "error_message": null,
                    "payment_method": "bank",
                    "payout_method_type": null,
                    "connector_transaction_id": "R3JCD5MCHK74STT5",
                    "cancellation_reason": null,
                    "unified_code": null,
                    "unified_message": null
                }
            ]
        },
        {
            "payout_id": "bd158643-6de0-4822-b5bc-4e528c5d2712",
            "merchant_id": "merchant_1712111780",
            "amount": 1,
            "currency": "EUR",
            "connector": "adyen",
            "payout_type": "card",
            "billing": null,
            "customer_id": "cus_NOWPAviaOF3z7Ynbi5MQ",
            "auto_fulfill": true,
            "email": "guest@example.com",
            "name": "John Doe",
            "phone": "999999999",
            "phone_country_code": "+65",
            "client_secret": null,
            "return_url": null,
            "business_country": null,
            "business_label": null,
            "description": "Its my first payout request",
            "entity_type": "Individual",
            "recurring": true,
            "metadata": {},
            "status": "success",
            "error_message": null,
            "error_code": null,
            "profile_id": "pro_sxXGxtmZC38hYU6MB624",
            "created": "2024-04-03T02:36:41.448Z",
            "attempts": [
                {
                    "attempt_id": "bd158643-6de0-4822-b5bc-4e528c5d2712_1",
                    "status": "success",
                    "amount": 1,
                    "currency": "EUR",
                    "connector": "adyen",
                    "error_code": null,
                    "error_message": null,
                    "payment_method": "card",
                    "payout_method_type": null,
                    "connector_transaction_id": "HX3B6P6QC8PQ7RT5",
                    "cancellation_reason": null,
                    "unified_code": null,
                    "unified_message": null
                }
            ]
        }
    ]
  }
</details>

- Returned response contains limited entries as specified using `limit`.
- `status` returned in the outer entity is valid
- `email`, `name` and `phone` are populated using customer's info

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by sahkal** [comment]
CI checks are failing can you fix them.

**Review by sahkal — CHANGES_REQUESTED** [review]
CI checks are failing can you please fix it!

**Review by Sakilmostak — DISMISSED** [review]

File: crates/api_models/src/payouts.rs [hunk 1]
```diff
@@ -437,8 +437,6 @@ pub struct PayoutAttemptResponse {
     pub connector_transaction_id: Option<String>,
     /// If the payout was cancelled the reason provided here
     pub cancellation_reason: Option<String>,
-    /// Provide a reference to a stored payout method
-    pub payout_token: Option<String>,
```
Comment (sahkal) [inline crates/api_models/src/payouts.rs:441]: why is this removed? Is this not being used

File: crates/router/src/core/payments/transformers.rs [hunk 2]
```diff
@@ -971,44 +971,57 @@ impl ForeignFrom<(storage::PaymentIntent, storage::PaymentAttempt)> for api::Pay
...
 #[cfg(feature = "payouts")]
-impl ForeignFrom<(storage::Payouts, storage::PayoutAttempt)> for api::PayoutCreateResponse {
-    fn foreign_from(item: (storage::Payouts, storage::PayoutAttempt)) -> Self {
+impl ForeignFrom<(storage::Payouts, storage::PayoutAttempt, domain::Customer)>
+    for api::PayoutCreateResponse
+{
+    fn foreign_from(item: (storage::Payouts, storage::PayoutAttempt, domain::Customer)) -> Self {
         let payout = item.0;
         let payout_attempt = item.1;
+        let customer = item.2;
```
Comment (sahkal) [inline crates/router/src/core/payments/transformers.rs:980]: can we do it something like 
```suggestion
        (payout, payout_attempt, customer) = item;
```

File: crates/router/src/core/payouts.rs [hunk 3]
```diff
@@ -684,7 +696,34 @@ pub async fn payouts_list_core(
...
+                {
+                    Ok(customer) => Some(Ok((payouts, payout_attempt, customer))),
+                    Err(error) => {
+                        if matches!(
+                            error.current_context(),
+                            storage_impl::errors::StorageError::ValueNotFound(_)
+                        ) {
+                            logger::warn!(
+                                ?error,
+                                "customer missing for customer_id : {}",
```
Comment (sahkal) [inline crates/router/src/core/payouts.rs:716]: Will it be wise to log errors with info suggesting its payout flow.

File: crates/router/src/core/payouts.rs [hunk 4]
```diff
@@ -738,8 +784,20 @@ pub async fn payouts_filtered_list_core(
         .await
         .to_not_found_response(errors::ApiErrorResponse::PayoutNotFound)?;
 
-    let data: Vec<api::PayoutCreateResponse> =
-        list.into_iter().map(ForeignFrom::foreign_from).collect();
+    let data: Vec<api::PayoutCreateResponse> = join_all(list.into_iter().map(|(p, pa, c)| async {
+        match domain::Customer::convert_back(c, &key_store.key).await {
+            Ok(domain_cust) => Some((p, pa, domain_cust)),
+            Err(err) => {
+                logger::warn!(?err, "failed to convert customer for id: {}", p.customer_id);
```
Comment (sahkal) [inline crates/router/src/core/payouts.rs:791]: +1

**Review by sahkal — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payouts.rs (hunk 1 above)
Comment (kashif-m) [reply crates/api_models/src/payouts.rs:441]: Yes, this is a temporary token which is not necessary for analytics

File: crates/router/src/core/payouts.rs (hunk 3 above)
Comment (kashif-m) [reply crates/router/src/core/payouts.rs:716]: Logs are already scoped with request IDs and flow type, right? We don't need to add more context about payouts in this log line if these are already fetched using request IDs

**Review by sahkal — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]
